agent/consul/state/catalog.go

```diff
@@ -4081,13 +4081,17 @@ func (s *Store) ServiceTopology(
 		maxIdx = idx
 	}
 
-	// Store downstreams with at least one instance in transparent proxy mode.
+	// Store downstreams with at least one instance in transparent proxy or connect native mode.
 	// This is to avoid returning downstreams from intentions when none of the downstreams are transparent proxies.
-	tproxyMap := make(map[structs.ServiceName]struct{})
+	proxyMap := make(map[structs.ServiceName]struct{})
 	for _, downstream := range unfilteredDownstreams {
 		if downstream.Service.Proxy.Mode == structs.ProxyModeTransparent {
 			sn := structs.NewServiceName(downstream.Service.Proxy.DestinationServiceName, &downstream.Service.EnterpriseMeta)
-			tproxyMap[sn] = struct{}{}
+			proxyMap[sn] = struct{}{}
+		}
+		if downstream.Service.Connect.Native {
+			sn := structs.NewServiceName(downstream.Service.Service, &downstream.Service.EnterpriseMeta)
```

Comment:
I am also not sure if `Service.Service` is 100% correct here, but it doesn't seem to fail tests? :D

Reply:
I'd recommend `downstream.Service.CompoundServiceName()` (which effectively does the same thing)